In [64]:
import numpy as np
import pandas as pd

นำเข้าข้อมูล

In [65]:
restaurant = pd.read_csv("Restaurant.csv")
restaurant

,Order ID,Date,Product,Price,Quantity,Purchase Type,Payment Method,Manager,City
0,10452,7/11/2022,Fries,3.49,573.07,Online,Gift Card,Tom Jackson,London
1,10453,7/11/2022,Beverages,2.95,745.76,Online,Gift Card,Pablo Perez,Madrid
2,10454,7/11/2022,Sides & Other,4.99,200.40,In-store,Gift Card,Joao Silva,Lisbon
3,10455,8/11/2022,Burgers,12.99,569.67,In-store,Credit Card,Walter Muller,Berlin
4,10456,8/11/2022,Chicken Sandwiches,9.95,201.01,In-store,Credit Card,Walter Muller,Berlin
...,...,...,...,...,...,...,...,...,...
249,10709,28-12-2022,Sides & Other,4.99,200.40,Drive-thru,Gift Card,Walter Muller,Berlin
250,10710,29-12-2022,Burgers,12.99,754.43,Drive-thru,Gift Card,Walter Muller,Berlin
251,10711,29-12-2022,Chicken Sandwiches,9.95,281.41,Drive-thru,Gift Card,Walter Muller,Berlin
252,10712,29-12-2022,Fries,3.49,630.37,Drive-thru,Gift Card,Walter Muller,Berlin


In [66]:
#ข้อมูลใน csv
restaurant.info()  #แสดงประเภทข้อมูล (Data Types) และจำนวน Non-Null   

<class 'pandas.DataFrame'>
RangeIndex: 254 entries, 0 to 253
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Order ID        254 non-null    int64  
 1   Date            254 non-null    str    
 2   Product         254 non-null    str    
 3   Price           254 non-null    float64
 4   Quantity        254 non-null    float64
 5   Purchase Type   254 non-null    str    
 6   Payment Method  254 non-null    str    
 7   Manager         254 non-null    str    
 8   City            254 non-null    str    
dtypes: float64(2), int64(1), str(6)
memory usage: 18.0 KB


จากผลลัพธ์การตรวจสอบข้อมูล แสดงให้เห็นว่า
1. ข้อมูลอยู่ในรูป dataframe
2. มีข้อมูลทั้งหมด 254 แถว 10 columns 
3. ข้อมูลแต่ละ Columns ไม่มีข้อมูลใดที่เป็นช่องว่าง
4. Date ถูกเก็บในรูปแบบ str ทำให้ไม่สามารถนำมากรองเป็นวันที่ได้
5. ค่า quantity ถูกเก็บเป็น float แต่จำนวนชิ้นในการขายควรเป็นจำนวณเต็ม

แปลง quantity

In [67]:
restaurant['Quantity'] = restaurant['Quantity'].round().astype(int)
restaurant.head()

,Order ID,Date,Product,Price,Quantity,Purchase Type,Payment Method,Manager,City
0,10452,7/11/2022,Fries,3.49,573,Online,Gift Card,Tom Jackson,London
1,10453,7/11/2022,Beverages,2.95,746,Online,Gift Card,Pablo Perez,Madrid
2,10454,7/11/2022,Sides & Other,4.99,200,In-store,Gift Card,Joao Silva,Lisbon
3,10455,8/11/2022,Burgers,12.99,570,In-store,Credit Card,Walter Muller,Berlin
4,10456,8/11/2022,Chicken Sandwiches,9.95,201,In-store,Credit Card,Walter Muller,Berlin


In [68]:
#ตรวจสอบรูปแบบของข้อมูล Date
restaurant['Date'].sample(10)
#พบว่ารูปแบบของวันที่ มีหลาายรูปแบบ ต้องเปลี่ยนให้เป็นรูปแบบเดียวกันเพื่อให้สามารถกรองข้อมูลวันที่ได้ในอนาคต

169    12/12/2022
231    25-12-2022
50     18-11-2022
98     28-11-2022
134     5/12/2022
219    22-12-2022
161    11/12/2022
184    15-12-2022
152     9/12/2022
39     16-11-2022
Name: Date, dtype: str

In [69]:
#Date column to datetime
#เปลี่ยนเครื่องหมาย / ให้เป็น - ทั้งหมด กรณีที่แปลงแล้ว โค้ดจะ Error เพราะหา / ไม่เจอ
restaurant['Date'] = restaurant['Date'].str.replace('/', '-')

In [70]:
#บันทึกไฟล์แยกเพื่อนับว่าไฟล์นี้เป็นไฟล์สำหรับทำงานต่อ เพื่อไม่ให้่ยุ่งเกี่ยวกับไฟล์ข้อมูลหลัก
restaurant.to_csv('C_Restaurant.csv', index=False)
print("save file success")

save file success


In [71]:
#ดึงไฟล์ที่เปลี่ยนเครื่องหมายแล้วมาใช้งานต่อ
CleanR = pd.read_csv("C_Restaurant.csv")
CleanR['Date'] = pd.to_datetime(restaurant['Date'], format='%d-%m-%Y')  #format='%d-%m-%Y' คือ บอกว่าข้อมูลถูกเรียงรูปแบบไหน
# เปลี่ยนรูปแบบข้อมูลของ date จาก str ให้เป็น Datetime
CleanR.info()

<class 'pandas.DataFrame'>
RangeIndex: 254 entries, 0 to 253
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Order ID        254 non-null    int64         
 1   Date            254 non-null    datetime64[us]
 2   Product         254 non-null    str           
 3   Price           254 non-null    float64       
 4   Quantity        254 non-null    int64         
 5   Purchase Type   254 non-null    str           
 6   Payment Method  254 non-null    str           
 7   Manager         254 non-null    str           
 8   City            254 non-null    str           
dtypes: datetime64[us](1), float64(1), int64(2), str(5)
memory usage: 18.0 KB


In [72]:
# สร้างฟ๊เจอร์แยกวัน เดือน และวันหยุด เพื่อวิเคราะห์ยอดขาายและพฤติกรรมของแต่ละช่วงเวลา  ไม่พิจารณาปี เนื่องจากข้อมูลที่มี มีเพียงสองเดือนของปี 2022
#แยกข้อมูลเดือน
CleanR['Month'] = CleanR['Date'].dt.month
#แยกวันที่ในเดือน
CleanR['Day'] = CleanR['Date'].dt.day
#แยกแต่ละวันในสัปดาห์ และต่อเนื่องไปถึงการวิเคราะห์ช่วงเวลาวันหยุด
CleanR['DayOfWeek'] = CleanR['Date'].dt.dayofweek  # 0=Monday, 6=Sunday
#ตรวจสอบวันหยุด เสาร์อาทิตย์ = 1, อื่น ๆ 0
CleanR['IsWeekend'] = CleanR['DayOfWeek'].isin([5, 6]).astype(int)

print(CleanR[['Date', 'Month', 'Day', 'DayOfWeek', 'IsWeekend']].sample(10))

          Date  Month  Day  DayOfWeek  IsWeekend
46  2022-11-17     11   17          3          0
45  2022-11-17     11   17          3          0
73  2022-11-23     11   23          2          0
26  2022-11-13     11   13          6          1
119 2022-12-02     12    2          4          0
234 2022-12-25     12   25          6          1
190 2022-12-17     12   17          5          1
150 2022-12-09     12    9          4          0
29  2022-11-14     11   14          0          0
77  2022-11-23     11   23          2          0


การแปรผลลัพธ์
วันที่่ 21 เดือน 12 เป็นวันพุธ ซึ่งไม่ใช่วันหยุด ค่า IsWeekend เป็น 0
วันที่่ 27 เดือน 11 เป็นวันอาทิตย์ ซึ่งเป็นวันหยุด ค่า IsWeekend เป็น 1

In [73]:
CleanR.head()

,Order ID,Date,Product,Price,Quantity,Purchase Type,Payment Method,Manager,City,Month,Day,DayOfWeek,IsWeekend
0,10452,2022-11-07,Fries,3.49,573,Online,Gift Card,Tom Jackson,London,11,7,0,0
1,10453,2022-11-07,Beverages,2.95,746,Online,Gift Card,Pablo Perez,Madrid,11,7,0,0
2,10454,2022-11-07,Sides & Other,4.99,200,In-store,Gift Card,Joao Silva,Lisbon,11,7,0,0
3,10455,2022-11-08,Burgers,12.99,570,In-store,Credit Card,Walter Muller,Berlin,11,8,1,0
4,10456,2022-11-08,Chicken Sandwiches,9.95,201,In-store,Credit Card,Walter Muller,Berlin,11,8,1,0


In [74]:
#ตรวจสอบ missing values
CleanR[CleanR.isnull().any(axis=1)].head()

,Order ID,Date,Product,Price,Quantity,Purchase Type,Payment Method,Manager,City,Month,Day,DayOfWeek,IsWeekend


In [75]:
#ตรวจหาข้อมูลซ้ำ
CleanR.duplicated().sum()

np.int64(0)

In [76]:
#ตรวจหาค่า Unique เพื่อตรวจสอบการไม่ซ้ำหรือ Null ของข้อมูลเพิ่มเติม
CleanR['Product'].unique()

<StringArray>
['Fries', 'Beverages', 'Sides & Other', 'Burgers', 'Chicken Sandwiches']
Length: 5, dtype: str

In [77]:
CleanR['Purchase Type'].unique()

<StringArray>
['Online ', 'In-store ', 'Drive-thru ']
Length: 3, dtype: str

In [78]:
CleanR['Payment Method'].unique()

<StringArray>
[' Gift Card', ' Credit Card', ' Cash']
Length: 3, dtype: str

In [79]:
CleanR['Manager'].unique()

<StringArray>
[  'Tom      Jackson', '       Pablo Perez',      'Joao    Silva',
      'Walter Muller',      'Remy    Monet',         'Remy Monet',
  '       Remy Monet',     'Remy     Monet',        'Pablo Perez',
      'Pablo   Perez',       'Pablo  Perez',     'Pablo    Perez',
         'Joao Silva',        'Tom Jackson']
Length: 14, dtype: str

In [80]:
CleanR['City'].unique()

<StringArray>
['London', 'Madrid', 'Lisbon', 'Berlin', 'Paris']
Length: 5, dtype: str

พบว่า Manager มีข้อมูลซ้ำ แต่นับว่า Unique เนื่องจากการเว้นวรรคช่องว่าง รวมถึง Purchase Type และ Payment Method มีช่องว่างที่ไม่จำเป็น ซึ่งอาจทำให้เกิดปัญหาเมื่อต้องใช้งาน ต้องจัดการกับช่องว่าง

In [81]:
# ใช้ Regex เพื่อยุบช่องว่างที่เกินกว่า 1 ช่องให้เหลือช่องว่างเดียว
text_columns = ['Purchase Type', 'Payment Method', 'Manager']
for col in text_columns:
   CleanR[col] = CleanR[col].str.replace(r'\s+', ' ', regex=True).str.strip()
# ตรวจสอบความถูกต้อง
print(CleanR['Purchase Type'].unique())
print(CleanR['Payment Method'].unique())
print(CleanR['Manager'].unique())

<StringArray>
['Online', 'In-store', 'Drive-thru']
Length: 3, dtype: str
<StringArray>
['Gift Card', 'Credit Card', 'Cash']
Length: 3, dtype: str
<StringArray>
['Tom Jackson', 'Pablo Perez', 'Joao Silva', 'Walter Muller', 'Remy Monet']
Length: 5, dtype: str


ตรวจสอบค่าทางสถิติ

In [82]:
CleanR.describe()

,Order ID,Date,Price,Quantity,Month,Day,DayOfWeek,IsWeekend
count,254.000000,254,254.000000,254.000000,254.000000,254.000000,254.000000,254.000000
mean,10584.133858,2022-12-03 10:23:37.322834,7.102323,460.480315,11.555118,16.779528,2.842520,0.251969
min,10452.000000,2022-11-07 00:00:00,2.950000,200.000000,11.000000,1.000000,0.000000,0.000000
25%,10520.250000,2022-11-21 00:00:00,3.490000,201.000000,11.000000,11.000000,1.000000,0.000000
50%,10583.500000,2022-12-03 00:00:00,4.990000,539.000000,12.000000,17.000000,3.000000,0.000000
75%,10649.750000,2022-12-16 18:00:00,9.950000,677.000000,12.000000,23.750000,4.750000,0.750000
max,10713.000000,2022-12-29 00:00:00,29.050000,754.000000,12.000000,30.000000,6.000000,1.000000
std,75.889181,NaN,4.341855,214.979431,0.497934,7.884957,1.955736,0.435000


ค่า Price มีการดีดสูงขึ้นผิดปกติ

In [83]:
CleanR.groupby('Product')['Price'].value_counts()

Product             Price
Beverages           2.95     50
Burgers             12.99    52
Chicken Sandwiches  9.95     51
                    29.05     1
Fries               3.49     50
                    25.50     1
Sides & Other       4.99     49
Name: count, dtype: int64

ราคา Chicken Sandwiches และ Fries มี ราคา ผิดปกติ 1 รายงาน แต่ สินค้าชื่อเดียวกันควรจะมีราคาต่อหน่วยเท่ากันทุกออเดอร์ ต้องแก้ไขราคาต่อหน่วยที่ผิดปกติ

In [84]:
standard_prices = { 'Beverages': 2.95,
                     'Burgers': 12.99,
                     'Chicken Sandwiches': 9.95,
                     'Fries': 3.49,
                     'Sides &amp; Other': 4.99
                   }

CleanR['Price'] = CleanR['Product'].map(standard_prices)

In [85]:
CleanR.groupby('Product')['Price'].value_counts()

Product             Price
Beverages           2.95     50
Burgers             12.99    52
Chicken Sandwiches  9.95     52
Fries               3.49     51
Name: count, dtype: int64

In [87]:
CleanR[['Price']].describe()

,Price
count,205.000000
mean,7.406683
std,4.277644
min,2.950000
25%,3.490000
50%,9.950000
75%,12.990000
max,12.990000


In [ ]:
เพิ่มข้อมูล Revenue ที่ต้องนำไปพิจารณา

In [89]:
CleanR['Revenue'] = CleanR['Price'] * CleanR['Quantity']
CleanR.sample(5)

,Order ID,Date,Product,Price,Quantity,Purchase Type,Payment Method,Manager,City,Month,Day,DayOfWeek,IsWeekend,Revenue
251,10711,2022-12-29,Chicken Sandwiches,9.95,281,Drive-thru,Gift Card,Walter Muller,Berlin,12,29,3,0,2795.95
246,10706,2022-12-28,Chicken Sandwiches,9.95,302,Drive-thru,Gift Card,Walter Muller,Berlin,12,28,2,0,3004.90
68,10525,2022-11-22,Burgers,12.99,462,Drive-thru,Cash,Remy Monet,Paris,11,22,1,0,6001.38
168,10628,2022-12-12,Beverages,2.95,678,Online,Gift Card,Joao Silva,Lisbon,12,12,0,0,2000.10
102,10562,2022-11-29,Fries,3.49,630,Online,Credit Card,Tom Jackson,London,11,29,1,0,2198.70


สรุปยอดขายตามประเภทสินค้า

In [91]:
product_summary = CleanR.groupby('Product').agg( Order_Count = ('Order ID', 'count'),
                                             Total_Quantity = ('Quantity', 'sum'),
                                             Price = ('Price', 'first'),
                                             Total_Revenue = ('Revenue', 'sum') ).reset_index()
product_summary

,Product,Order_Count,Total_Quantity,Price,Total_Revenue
0,Beverages,50,34988,2.95,103214.60
1,Burgers,52,29018,12.99,376943.82
2,Chicken Sandwiches,52,11133,9.95,110773.35
3,Fries,51,32023,3.49,111760.27
4,Sides & Other,49,9800,NaN,0.00


In [ ]:
คำนวณ % สัดส่วนยอดขาย

In [95]:
product_summary['Revenue_Share (%)'] = (product_summary['Total_Revenue'] / 
                                        product_summary['Total_Revenue'].sum()) * 100
product_summary['Revenue_Share (%)']

0    14.688454
1    53.642819
2    15.764139
3    15.904587
4     0.000000
Name: Revenue_Share (%), dtype: float64

In [ ]:
จัดเรียงจากยอดขายสูงสุดไปต่ำสุด

In [97]:
product_summary = product_summary.sort_values(by='Total_Revenue',
                                              ascending=False)
display(product_summary)

,Product,Order_Count,Total_Quantity,Price,Total_Revenue,Revenue_Share (%)
1,Burgers,52,29018,12.99,376943.82,53.642819
3,Fries,51,32023,3.49,111760.27,15.904587
2,Chicken Sandwiches,52,11133,9.95,110773.35,15.764139
0,Beverages,50,34988,2.95,103214.60,14.688454
4,Sides & Other,49,9800,NaN,0.00,0.000000


ตรวจสอบค่าว่าง ค่าซ้ำ อีกครั้ง

In [98]:
print(CleanR.duplicated().sum())

0
